src/server/pps/server/api_server.go

```diff
@@ -1519,19 +1448,197 @@ func (a *apiServer) getLogsFromStats(pachClient *client.APIClient, request *pps.
 					continue
 				}
 
-				mu.Lock()
-				if err := apiGetLogsServer.Send(msg); err != nil {
-					mu.Unlock()
-					return err
+				select {
+				case logCh <- msg:
+				case <-ctx.Done():
+					return nil
 				}
-				mu.Unlock()
 			}
 			return nil
 		})
 	}
 	return eg.Wait()
 }
 
+func (a *apiServer) followLogsForRC(pachClient *client.APIClient, request *pps.GetLogsRequest, containerName, rcName string, logCh chan *pps.LogMessage) error {
+	defer close(logCh)
+	ctx := pachClient.Ctx()
+
+	err := backoff.RetryNotify(func() error {
+		pods, err := a.rcPodsFromLogsRequest(pachClient, request, rcName)
+		if err != nil {
+			return err
+		}
+
+		// We'll spawn a separate goro for each pod, which will all be
+		// managed by this error group. The behavior of an error group's
+		// `.Wait()` (called below) is to return when all of the goros have
+		// returned. By using a context tied to the error group, a single
+		// failing goro can signal to the other goros to wrap up as well.
+		// Because the error group context builds upon the request context,
+		// it can also be cancelled when the request is cancelled.
+		eg, egCtx := errgroup.WithContext(ctx)
+
+		for _, pod := range pods {
+			// Without this, variable shadowing will prevent us from
+			// properly iterating over all the pods
+			pod := pod
+
+			eg.Go(func() error {
+				return a.getLogsFromPod(egCtx, request, pod, containerName, logCh)
+			})
+		}
+
+		return eg.Wait()
+	}, backoff.NewConstantBackOff(5*time.Second), func(err error, _ time.Duration) error {
+		// Keep retrying, unless the request context is done
+		select {
+		case <-ctx.Done():
+			return err
+		default:
+			return nil
+		}
+	})
+
+	// Do not propagate EOF errors, as they're strictly used to signal
+	// shutdown for pod log goros
```

Comment:
I'm afraid I don't understand this part. Why return EOF from `getLogsFromPod` and then ignore it here?

Reply:
When using error groups with contexts, the context will close as soon as the first worker returns an error. I'm using an `io.EOF` here to signal that one of the workers finished successfully, so that the other workers get notified to close.

As an aside, I debated whether to create a custom error or use `io.EOF`. I think the latter is right because:

1) It's in keeping with go stdlib's use of EOF as a signal of a valid state rather than an error.
2) In a sense, an EOF did happen, as the underlying log stream was closed.

Reply:
Got it, thanks! I guess my only remaining question is: do we want to close all per-pod goros as soon as the first goro finishes? It seems like that means we'll drop logs from all goros after the first—is that what happens/is that actually the desired behavior?

Reply:
That's a good point, I'm not sure. I thought it was the right way because it protects against leaks (e.g. one log stream somehow staying open even when the others are winding down), but it likely introduces its own problems. Luckily, this'll be moot once I switch this over to getting service logs.